README.md
# modules-api
E-Com Plus Mods API Node.js App

[REST API reference](https://developers.e-com.plus/modules-api/)

# Technology stack
+ [NodeJS](https://nodejs.org/en/) 12.x
+ [REST Auto Router](https://github.com/leomp12/nodejs-rest-auto-router) npm package

# Setting up
```bash
git clone https://github.com/ecomclub/modules-api
cd modules-api
cp config/config-sample.json config/config.json
nano config/config.json
```

Edit `config.json` placing correct values for your environment,
after that, start app with node:

```bash
node ./main.js
```

# Web server
You need to use a web server such as NGINX or Apache HTTP,
proxy passing the requests to configured TCP port.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.12.52](https://github.com/ecomplus/modules-api/compare/v0.12.51...v0.12.52) (2022-07-28)


### Features

* **apply-discount:** add `buy_together` (optional) to apply discount response schema ([#17](https://github.com/ecomplus/modules-api/issues/17)) ([f328ff6](https://github.com/ecomplus/modules-api/commit/f328ff658febb8acf8a0b60df0317e914d242552))

### [0.12.51](https://github.com/ecomplus/modules-api/compare/v0.12.50...v0.12.51) (2022-05-24)


### Bug Fixes

* **checkout:** prevent amount being changed by reference on secondary transactions ([c5ff18e](https://github.com/ecomplus/modules-api/commit/c5ff18ef4c82d50795c2b10157fe4a0d9bf8f52d))

### [0.12.50](https://github.com/ecomplus/modules-api/compare/v0.12.49...v0.12.50) (2022-05-24)


### Bug Fixes

* **checkout:** must set payment history entriy id ([73cab7b](https://github.com/ecomplus/modules-api/commit/73cab7b7f64f15c4708ab3a35c9d1b85b6599b07))

### [0.12.49](https://github.com/ecomplus/modules-api/compare/v0.12.48...v0.12.49) (2022-05-24)


### Features

* **checkout:** save loyalty points amount as `amount.balance` on new order ([cfc78e3](https://github.com/ecomplus/modules-api/commit/cfc78e396cf0eb9503d3ff951742d42f4596e707))

### [0.12.48](https://github.com/ecomplus/modules-api/compare/v0.12.47...v0.12.48) (2022-04-21)


### Bug Fixes

* **checkout:** also accept ipv6 `browser_ip` ([856702c](https://github.com/ecomplus/modules-api/commit/856702c5ee4508c9d217db83386053b3d6ffce67))

### [0.12.47](https://github.com/ecomplus/modules-api/compare/v0.12.46...v0.12.47) (2022-04-21)


### Bug Fixes

* **checkout:** prevent error with invalid client ip address ([c768c44](https://github.com/ecomplus/modules-api/commit/c768c441d366f12647bab8ab620481c0f98ec0ef))

### [0.12.46](https://github.com/ecomplus/modules-api/compare/v0.12.45...v0.12.46) (2022-04-21)


### Bug Fixes

* **checkout:** set `browser_ip` (if undefined) with request client ip address ([f7ffecf](https://github.com/ecomplus/modules-api/commit/f7ffecf7a5ebb11e2c122a297090eb133d1a1b44))

### [0.12.45](https://github.com/ecomplus/modules-api/compare/v0.12.44...v0.12.45) (2022-01-24)


### Bug Fixes

* **checkout:** return invalid items when no subtotal and some item removed ([9caf043](https://github.com/ecomplus/modules-api/commit/9caf043ae26d1c0eba7d2d4542861f5cabebbed7))

### [0.12.44](https://github.com/ecomplus/modules-api/compare/v0.12.43...v0.12.44) (2021-12-27)


### Bug Fixes

* **checkout:** better object id (for item ids) function also using `Math.random` ([b2df539](https://github.com/ecomplus/modules-api/commit/b2df539bde0c2edd060a35138dfd8e9c0af6077b))
* **checkout:** properly complete `transaction.app` object when multiple transactions ([688f120](https://github.com/ecomplus/modules-api/commit/688f120f04a8d4ab53c8f06c2fd35474980b5870))

### [0.12.43](https://github.com/ecomplus/modules-api/compare/v0.12.42...v0.12.43) (2021-10-26)


### Bug Fixes

* **applications:** prevent aborting requests by queue delay when multiple apps ([0ff9da2](https://github.com/ecomplus/modules-api/commit/0ff9da236874782f92d96fd757d2e49ea7feb801))

### [0.12.42](https://github.com/ecomplus/modules-api/compare/v0.12.41...v0.12.42) (2021-10-26)


### Bug Fixes

* **modules:** prevent error with undefined `response` object ([b94a0dd](https://github.com/ecomplus/modules-api/commit/b94a0dd4a66775e5c6df3358a6fe4013d9913475))

### [0.12.41](https://github.com/ecomplus/modules-api/compare/v0.12.40...v0.12.41) (2021-10-26)


### Bug Fixes

* **modules:** should restart on 50+ consecutive req aborts ([4a1f674](https://github.com/ecomplus/modules-api/commit/4a1f67478f8fc56cd6ff2f985a05918a94172835))

### [0.12.40](https://github.com/ecomplus/modules-api/compare/v0.12.39...v0.12.40) (2021-10-26)


### Bug Fixes

* **applications:** respond with status 429 instead of 503 on big queue delays ([c003eb0](https://github.com/ecomplus/modules-api/commit/c003eb0ccbec3cec2acb5a0b5bab9396f4d63214))

### [0.12.39](https://github.com/ecomplus/modules-api/compare/v0.12.38...v0.12.39) (2021-10-22)


### Bug Fixes

* **checkout:** also set order `financial_status` after first transactions response ([71faac3](https://github.com/ecomplus/modules-api/commit/71faac3884b1bdf3d301b956c04a697f0ba760b0))


routes/sh.js
'use strict'

// import common verbs (methods) functions
const httpVerbs = require('./#modules.js')
const modName = 'sh'

const schema = {
  '$schema': 'http://json-schema.org/draft-06/schema#',
  'title': 'Module SH: Input model',
  'description': 'On shipping freight calculation',
  'type': 'object',
  'required': [ 'items', 'from', 'to' ],
  'additionalProperties': false,
  'definitions': {
    'address': {
      'type': 'object',
      'additionalProperties': false,
      'required': [ 'zip' ],
      'properties': {
        'zip': {
          'type': 'string',
          'maxLength': 30,
          'description': 'ZIP (CEP, postal...) code'
        },
        'street': {
          'type': 'string',
          'maxLength': 200,
          'description': 'Street or public place name'
        },
        'number': {
          'type': 'integer',
          'min': 1,
          'max': 9999999,
          'description': 'House or building street number'
        },
        'complement': {
          'type': 'string',
          'maxLength': 100,
          'description': 'Address complement or second line, such as apartment number'
        },
        'borough': {
          'type': 'string',
          'maxLength': 100,
          'description': 'Borough name'
        },
        'near_to': {
          'type': 'string',
          'maxLength': 100,
          'description': 'Some optional other reference for this address'
        },
        'line_address': {
          'type': 'string',
          'maxLength': 255,
          'description': 'Full in line mailing address, should include street, number and borough'
        },
        'city': {
          'type': 'string',
          'maxLength': 100,
          'description': 'City name'
        },
        'country': {
          'type': 'string',
          'maxLength': 50,
          'description': 'Country name'
        },
        'country_code': {
          'type': 'string',
          'minLength': 2,
          'maxLength': 2,
          'pattern': '^[A-Z]+$',
          'description': 'An ISO 3166-2 country code'
        },
        'province': {
          'type': 'string',
          'maxLength': 100,
          'description': 'Province or state name'
        },
        'province_code': {
          'type': 'string',
          'minLength': 2,
          'maxLength': 2,
          'pattern': '^[A-Z]+$',
          'description': 'The two-letter code for the province or state'
        },
        'name': {
          'type': 'string',
          'maxLength': 70,
          'description': 'The name of recipient, generally is the customer\'s name'
        },
        'last_name': {
          'type': 'string',
          'maxLength': 70,
          'description': 'The recipient\'s last name'
        },
        'phone': {
          'type': 'object',
          'additionalProperties': false,
          'required': [ 'number' ],
          'properties': {
            'country_code': {
              'type': 'integer',
              'min': 1,
              'max': 999,
              'description': 'Country calling code (without +), defined by standards E.123 and E.164'
            },
            'number': {
              'type': 'string',
              'maxLength': 19,
              'pattern': '^[0-9]+$',
              'description': 'The actual phone number, digits only'
            }
          },
          'description': 'Customer phone number for this mailing address'
        }
      },
      'description': 'Address object'
    }
  },
  'properties': {
    'items': {
      'type': 'array',
      'maxItems': 3000,
      'items': {
        'type': 'object',
        'additionalProperties': false,
        'required': [ 'product_id', 'quantity', 'price' ],
        'properties': {
          'product_id': {
            'type': 'string',
            'pattern': '^[a-f0-9]{24}$',
            'description': 'Product ID'
          },
          'variation_id': {
            'type': 'string',
            'pattern': '^[a-f0-9]{24}$',
            'description': 'ID to specify the variation added to cart, if product has variations'
          },
          'sku': {
            'type': 'string',
            'minLength': 2,
            'maxLength': 100,
            'pattern': '^[A-Za-z0-9-_.]+$',
            'description': 'Product or variation unique reference code'
          },
          'name': {
            'type': 'string',
            'maxLength': 255,
            'description': 'Product or variation full name, or other label for this cart item'
          },
          'quantity': {
            'type': 'number',
            'multipleOf': 0.0001,
            'minimum': 0,
            'maximum': 9999999,
            'description': 'Item quantity in cart'
          },
          'currency_id': {
            'type': 'string',
            'pattern': '^[A-Z]{3}$',
            'default': 'BRL',
            'description': 'Designator of currency according to ISO 4217 (3 uppercase letters)'
          },
          'currency_symbol': {
            'type': 'string',
            'maxLength': 20,
            'default': 'R$',
            'description': 'Graphic symbol used as a shorthand for currency\'s name'
          },
          'price': {
            'type': 'number',
            'multipleOf': 0.00001,
            'minimum': 0,
            'maximum': 999999999,
            'description': 'Product sale price specifically for this cart'
          },
          'final_price': {
            'type': 'number',
            'multipleOf': 0.00001,
            'minimum': 0,
            'maximum': 999999999,
            'description': 'Final item price including additions due to customizations and gift wrap'
          },
          'dimensions': {
            'type': 'object',
            'additionalProperties': false,
            'patternProperties': {
              '^width|height|length$': {
                'type': 'object',
                'required': [ 'value' ],
                'additionalProperties': false,
                'properties': {
                  'value': {
                    'type': 'number',
                    'multipleOf': 0.001,
                    'minimum': 0,
                    'maximum': 999999,
                    'description': 'Size in specified unit'
                  },
                  'unit': {
                    'type': 'string',
                    'enum': [ 'mm', 'cm', 'm', 'ft', 'in', 'yd', 'mi' ],
                    'default': 'cm',
                    'description': 'Unit of measurement'
                  }
                },
                'description': 'Package width, height and length'
              }
            },
            'description': 'Product dimensions with packaging for freight calculation'
          },
          'weight': {
            'type': 'object',
            'required': [ 'value' ],
            'additionalProperties': false,
            'properties': {
              'value': {
                'type': 'number',
                'multipleOf': 0.001,
                'minimum': 0,
                'maximum': 999999,
                'description': 'Size in specified unit'
              },
              'unit': {
                'type': 'string',
                'enum': [ 'mg', 'g', 'kg', 'lb', 'oz' ],
                'default': 'g',
                'description': 'Unit of measurement'
              }
            },
            'description': 'Package weight for freight calculation'
          }
        },
        'description': 'One of the cart items'
      },
      'description': 'Products composing the cart'
    },
    'from': {
      '$ref': '#/definitions/address',
      'description': 'Sender\'s address'
    },
    'to': {
      '$ref': '#/definitions/address',
      'description': 'Shipping address (recipient)'
    },
    'own_hand': {
      'type': 'boolean',
      'default': false,
      'description': 'Whether the package must be delivered with additional service "own hand"'
    },
    'receipt': {
      'type': 'boolean',
      'default': false,
      'description': 'If the package will be delivered with acknowledgment of receipt'
    },
    'service_code': {
      'type': 'string',
      'maxLength': 70,
      'description': 'Code of service defined by carrier, if shipping method is already defined'
    }
  }
}

// validate module packages responses
const responseSchema = {
  '$schema': 'http://json-schema.org/draft-06/schema#',
  'title': 'Module SH: Package response model',
  'description': schema.description,
  'type': 'object',
  'required': [ 'shipping_services' ],
  'additionalProperties': false,
  'definitions': {
    ...schema.definitions
  },
  'properties': {
    'shipping_services': {
      'type': 'array',
      'minItems': 1,
      'maxItems': 30,
      'items': {
        'type': 'object',
        'additionalProperties': false,
        'required': [ 'label', 'shipping_line' ],
        'properties': {
          'label': {
            'type': 'string',
            'maxLength': 50,
            'description': 'Name of shipping method shown to customers'
          },
          'carrier': {
            'type': 'string',
            'maxLength': 200,
            'description': 'Carrier name'
          },
          'carrier_doc_number': {
            'type': 'string',
            'maxLength': 19,
            'pattern': '^[0-9]+$',
            'description': 'Carrier document number (only numbers)'
          },
          'service_name': {
            'type': 'string',
            'maxLength': 70,
            'description': 'Name of service (shipping method) defined by carrier'
          },
          'service_code': {
            'type': 'string',
            'maxLength': 70,
            'description': 'Code of service defined by carrier'
          },
          'contract': {
            'type': 'string',
            'maxLength': 70,
            'description': 'Contract number of the shop with carrier, if exists'
          },
          'icon': {
            'type': 'string',
            'maxLength': 255,
            'format': 'uri',
            'description': 'Shipping method icon image URI'
          },
          'rule': {
            'type': 'object',
            'required': [ '_id' ],
            'additionalProperties': false,
            'properties': {
              '_id': {
                'type': 'string',
                'pattern': '^[a-f0-9]{24}$',
                'description': 'Rule ID'
              },
              'label': {
                'type': 'string',
                'maxLength': 50,
                'description': 'Shipping rule label'
              }
            },
            'description': 'Selected shipping rule of this shipping method'
          },
          'shipping_line': {
            'type': 'object',
            'additionalProperties': false,
            'required': [ 'from', 'to' ],
            'properties': {
              'from': {
                '$ref': '#/definitions/address',
                'description': 'Sender\'s address'
              },
              'to': {
                '$ref': '#/definitions/address',
                'description': 'Shipping address (recipient)'
              },
              'package': {
                'type': 'object',
                'additionalProperties': false,
                'properties': {
                  'dimensions': {
                    'type': 'object',
                    'additionalProperties': false,
                    'patternProperties': {
                      '^width|height|length$': {
                        'type': 'object',
                        'required': [ 'value' ],
                        'additionalProperties': false,
                        'properties': {
                          'value': {
                            'type': 'number',
                            'multipleOf': 0.001,
                            'minimum': 0,
                            'maximum': 999999,
                            'description': 'Size in specified unit'
                          },
                          'unit': {
                            'type': 'string',
                            'enum': [ 'mm', 'cm', 'm', 'ft', 'in', 'yd', 'mi' ],
                            'default': 'cm',
                            'description': 'Unit of measurement'
                          }
                        },
                        'description': 'Package width, height and length'
                      }
                    },
                    'description': 'Package dimensions'
                  },
                  'weight': {
                    'type': 'object',
                    'required': [ 'value' ],
                    'additionalProperties': false,
                    'properties': {
                      'value': {
                        'type': 'number',
                        'multipleOf': 0.001,
                        'minimum': 0,
                        'maximum': 999999,
                        'description': 'Size in specified unit'
                      },
                      'unit': {
                        'type': 'string',
                        'enum': [ 'mg', 'g', 'kg', 'lb', 'oz' ],
                        'default': 'g',
                        'description': 'Unit of measurement'
                      }
                    },
                    'description': 'Package weight for freight calculation'
                  }
                },
                'description': 'Shipping object information'
              },
              'price': {
                'type': 'number',
                'multipleOf': 0.00001,
                'minimum': 0,
                'maximum': 9999999999,
                'description': 'Freight cost for this shipping line, without additionals'
              },
              'declared_value': {
                'type': 'number',
                'multipleOf': 0.00001,
                'minimum': 0,
                'maximum': 9999999999,
                'description': 'The package value declared to carrier, generally the sum of all items prices'
              },
              'declared_value_price': {
                'type': 'number',
                'multipleOf': 0.00001,
                'minimum': 0,
                'maximum': 999999,
                'description': 'Cost for the "declared value" additional service'
              },
              'own_hand': {
                'type': 'boolean',
                'default': false,
                'description': 'Whether the package must be delivered with additional service "own hand"'
              },
              'own_hand_price': {
                'type': 'number',
                'multipleOf': 0.00001,
                'minimum': 0,
                'maximum': 999999,
                'description': 'Cost for the "own hand" additional service'
              },
              'receipt': {
                'type': 'boolean',
                'default': false,
                'description': 'If the package will be delivered with acknowledgment of receipt'
              },
              'receipt_price': {
                'type': 'number',
                'multipleOf': 0.00001,
                'minimum': 0,
                'maximum': 999999,
                'description': 'Cost for the "acknowledgment of receipt" additional service'
              },
              'other_additionals': {
                'type': 'array',
                'maxItems': 30,
                'items': {
                  'type': 'object',
                  'additionalProperties': false,
                  'required': [ 'label' ],
                  'properties': {
                    'tag': {
                      'type': 'string',
                      'maxLength': 20,
                      'pattern': '^[a-z0-9_]+$',
                      'description': 'Tag to identify object, use only lowercase letters, digits and underscore'
                    },
                    'label': {
                      'type': 'string',
                      'maxLength': 50,
                      'description': 'Name of the additional service'
                    },
                    'price': {
                      'type': 'number',
                      'multipleOf': 0.00001,
                      'minimum': 0,
                      'maximum': 999999,
                      'description': 'Cost for this additional service'
                    }
                  },
                  'description': 'Additional service'
                },
                'description': 'List of other additional services for this shipping line'
              },
              'taxes': {
                'type': 'array',
                'maxItems': 30,
                'items': {
                  'type': 'object',
                  'additionalProperties': false,
                  'required': [ 'label' ],
                  'properties': {
                    'tag': {
                      'type': 'string',
                      'maxLength': 20,
                      'pattern': '^[a-z0-9_]+$',
                      'description': 'Tag to identify object, use only lowercase letters, digits and underscore'
                    },
                    'label': {
                      'type': 'string',
                      'maxLength': 50,
                      'description': 'Tax title'
                    },
                    'price': {
                      'type': 'number',
                      'multipleOf': 0.00001,
                      'minimum': 0,
                      'maximum': 999999999,
                      'description': 'Tax value applied'
                    },
                    'rate': {
                      'type': 'number',
                      'multipleOf': 0.0001,
                      'minimum': 0,
                      'maximum': 100,
                      'description': 'Tax rate as a function of package value'
                    }
                  },
                  'description': 'Applied tax or additional service'
                },
                'description': 'List of taxes or other additional services for this shipping line'
              },
              'discount': {
                'type': 'number',
                'multipleOf': 0.0001,
                'minimum': -999999999,
                'maximum': 999999999,
                'description': 'Discount on shipping price, negative if value was additionated (not discounted)'
              },
              'total_price': {
                'type': 'number',
                'multipleOf': 0.00001,
                'minimum': 0,
                'maximum': 9999999999,
                'description': 'Total cost for this shipping line, with additionals and taxes'
              },
              'posting_deadline': {
                'type': 'object',
                'required': [ 'days' ],
                'additionalProperties': false,
                'properties': {
                  'days': {
                    'type': 'integer',
                    'minimum': 0,
                    'maximum': 999999,
                    'description': 'Number of days to post the product after purchase'
                  },
                  'working_days': {
                    'type': 'boolean',
                    'default': true,
                    'description': 'If the deadline is calculated in working days'
                  },
                  'after_approval': {
                    'type': 'boolean',
                    'default': true,
                    'description': 'Whether days will be counted after payment approval'
                  }
                },
                'description': 'Deadline for sending the package'
              },
              'delivery_time': {
                'type': 'object',
                'required': [ 'days' ],
                'additionalProperties': false,
                'properties': {
                  'days': {
                    'type': 'integer',
                    'minimum': 0,
                    'maximum': 999999,
                    'description': 'Number of days for delivery after shipping'
                  },
                  'working_days': {
                    'type': 'boolean',
                    'default': true,
                    'description': 'If the deadline is calculated in working days'
                  }
                },
                'description': 'Estimated delivery time'
              },
              'scheduled_delivery': {
                'type': 'string',
                'format': 'date-time',
                'description': 'When delivery will be made on the ISO 8601 date time representation'
              },
              'flags': {
                'type': 'array',
                'uniqueItems': true,
                'maxItems': 10,
                'items': {
                  'type': 'string',
                  'maxLength': 20,
                  'description': 'Flag title'
                },
                'description': 'Flags to associate additional info'
              },
              'custom_fields': {
                'type': 'array',
                'maxItems': 10,
                'items': {
                  'type': 'object',
                  'additionalProperties': false,
                  'required': [ 'field', 'value' ],
                  'properties': {
                    'field': {
                      'type': 'string',
                      'maxLength': 50,
                      'description': 'Field name'
                    },
                    'value': {
                      'type': 'string',
                      'maxLength': 255,
                      'description': 'Field value'
                    }
                  },
                  'description': 'Custom field object'
                },
                'description': 'List of custom fields'
              },
              'notes': {
                'type': 'string',
                'maxLength': 255,
                'description': 'Optional notes with additional info about this shipping line'
              }
            },
            'description': 'Order shipping line object'
          }
        },
        'description': 'Shipping method option (service)'
      },
      'description': 'Shipping method options list'
    }
  }
}

module.exports = httpVerbs(modName, schema, responseSchema)
